### PR TITLE
Add KubeEdge integration proposal

### PR DESCRIPTION
**Pull request by Ma-Dan**

### CODE REVIEW

File: sig-edge/concepts-and-designs/kubesphere-kubeedge-integration.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+- Convenient edge node joining and exiting
+
+# E1. Enable and configure KubeEdge CloudCore via UI
+Users can toggle enable/disable and modify parameters of KubeEdge CloudClore on KubeSphere UI. Each cluster can config its KubeEdge component separately, only users with cluster-admin role can config KubeEdge. Configure parameters are shown below:
+
+## CloudCore config
+
+| Parameter | Description | Type | Default |
+| --------- | ----------- | ---- | ------- |
+| HA deploy | HA deployment of CloudCore | Bool | False |
```
Comment (benjaminhuo) [inline sig-edge/concepts-and-designs/kubesphere-kubeedge-integration.md:18]: HA deploy => CloudCore HA ?

File: sig-edge/concepts-and-designs/kubesphere-kubeedge-integration.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+| Service Type | Service expose methon | String | NodePort (NodePort/LoadBalancer) |
+| CloudHub advertiseAddress | Exposed IP address for EdgeCore to connect | String | Public IP address of KS cluster in NodePort mode / Service attached LB IP |
+| CloudHub nodeLimit | node limit for edge nodes | Int | 10 |
```
Comment (benjaminhuo) [inline sig-edge/concepts-and-designs/kubesphere-kubeedge-integration.md:21]: This means the upper limit of edge nodes that could be managed by this cloud hub?
Maybe the default 10 is too small? 100?

File: sig-edge/concepts-and-designs/kubesphere-kubeedge-integration.md [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+| CloudStream enable | Enable CloudStream component, required by logs and metrics | Bool | True |
```
Comment (benjaminhuo) [inline sig-edge/concepts-and-designs/kubesphere-kubeedge-integration.md:22]: CloudStream enable => Enable CloudStream?
Param for cloudstream should be put together.
Same for CloudHub and ClouldTunnel

File: sig-edge/concepts-and-designs/kubesphere-kubeedge-integration.md [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+| CloudHub https port | Port number of CloudHub https service | Int | 31002(NodePort) / 10002(LB) |
+| CloudStream port  | Port number of CloudHub service | Int | 31003(NodePort) / 10003(LB) |
+| CloudTunnel port  | Port number of CloudHub service | Int | 31004(NodePort) / 10004(LB) |
+
+- KubeEdge component can be installed by one helm chart.
+- After KubeEdge enabled, UI should prompt users to enable firewall and port mapping for KubeEdge service ports.
+
+# E2. Edge nodes logging and metrics
+Edge node reported IP is not an actual IP address which could be accessed by k8s apiserver and metrics-server in k8s cluster, and edge nodes can not be connected from k8s through Internet, calling to logs/metrics API must be forwarded by CloudStream service on top of the connection created from edge nodes to CloudCore, the data chain is like following:
+k8s apiserver -> edgenode-reported-ip:10250 -> CloudCore:10003 | CloudStream:10004 <- edgenode metrics server -> edgenode:10350
```
Comment (benjaminhuo) [inline sig-edge/concepts-and-designs/kubesphere-kubeedge-integration.md:33]: <- edgenode metrics server -> edgenode:10350
should be changed to
<- edgenode log endpoint -> edgenode:10250
?

**Review by benjaminhuo — CHANGES_REQUESTED** [review]

File: sig-edge/concepts-and-designs/kubesphere-kubeedge-integration.md [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+Each edge node required 2 iptables entries to enable logging and metrics fetching.
+TCP 192.168.100.1:10250 -> cloudcore-ip:10003(Can be 192.168.1.10:31003 in NodePort mode)
+TCP 192.168.100.1:10350 -> cloudcore-ip:10003(Can be 192.168.1.10:31003 in NodePort mode)
+
+Edge-router controller component is developed to maintain iptables entries according to edge nodes status, and installs together with kubeedge when enables it.
+
+![Edge-router](../images/edge-router.png)
+
+KubeSphere fetches edge nodes' real-time metrics values using same APIs like kubectl top.
```
Comment (benjaminhuo) [inline sig-edge/concepts-and-designs/kubesphere-kubeedge-integration.md:44]: Maybe it's better to list all available metrics here such as node, pods and container metrics

**Review by benjaminhuo — CHANGES_REQUESTED** [review]

File: sig-edge/concepts-and-designs/kubesphere-kubeedge-integration.md [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+# E1. Enable and configure KubeEdge CloudCore via UI
+Users can toggle enable/disable and modify parameters of KubeEdge CloudClore on KubeSphere UI. Each cluster can config its KubeEdge component separately, only users with cluster-admin role can config KubeEdge. Configure parameters are shown below:
+
+## CloudCore config
+
+| Parameter | Description | Type | Default |
+| --------- | ----------- | ---- | ------- |
+| HA deploy | HA deployment of CloudCore | Bool | False |
+| Service Type | Service expose methon | String | NodePort (NodePort/LoadBalancer) |
```
[the remark this answers is not in the document]
Comment (Ma-Dan) [reply sig-edge/concepts-and-designs/kubesphere-kubeedge-integration.md]: Fixed

File: sig-edge/concepts-and-designs/kubesphere-kubeedge-integration.md [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+| CloudHub advertiseAddress | Exposed IP address for EdgeCore to connect | String | Public IP address of KS cluster in NodePort mode / Service attached LB IP |
+| CloudHub nodeLimit | node limit for edge nodes | Int | 10 |
+| CloudStream enable | Enable CloudStream component, required by logs and metrics | Bool | True |
+| CloudHub port | Port number of CloudHub service | Int | 31000(NodePort) / 10000(LB) |
+| CloudHub https port | Port number of CloudHub https service | Int | 31002(NodePort) / 10002(LB) |
+| CloudStream port  | Port number of CloudHub service | Int | 31003(NodePort) / 10003(LB) |
```
[the remark this answers is not in the document]
Comment (Ma-Dan) [reply sig-edge/concepts-and-designs/kubesphere-kubeedge-integration.md]: Fixed

File: sig-edge/concepts-and-designs/kubesphere-kubeedge-integration.md [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+| CloudTunnel port  | Port number of CloudHub service | Int | 31004(NodePort) / 10004(LB) |
```
[the remark this answers is not in the document]
Comment (Ma-Dan) [reply sig-edge/concepts-and-designs/kubesphere-kubeedge-integration.md]: Fixed

File: sig-edge/concepts-and-designs/kubesphere-kubeedge-integration.md (hunk 1 above)
Comment (Ma-Dan) [reply sig-edge/concepts-and-designs/kubesphere-kubeedge-integration.md]: Fixed

File: sig-edge/concepts-and-designs/kubesphere-kubeedge-integration.md (hunk 2 above)
Comment (Ma-Dan) [reply sig-edge/concepts-and-designs/kubesphere-kubeedge-integration.md]: Fixed

File: sig-edge/concepts-and-designs/kubesphere-kubeedge-integration.md (hunk 3 above)
Comment (Ma-Dan) [reply sig-edge/concepts-and-designs/kubesphere-kubeedge-integration.md]: Fixed

File: sig-edge/concepts-and-designs/kubesphere-kubeedge-integration.md (hunk 4 above)
Comment (Ma-Dan) [reply sig-edge/concepts-and-designs/kubesphere-kubeedge-integration.md]: Fixed

File: sig-edge/concepts-and-designs/kubesphere-kubeedge-integration.md (hunk 5 above)
Comment (Ma-Dan) [reply sig-edge/concepts-and-designs/kubesphere-kubeedge-integration.md:57]: Added metrics detail

**Review by benjaminhuo — APPROVED** [review]

**Pull request merged**